packages/router/utils.ts

```diff
@@ -415,48 +415,58 @@ function flattenRoutes<
       return;
     }
 
-    // Handle optional params - /path/:optional?
-    let segments = path.split("/");
-    let optionalParams: string[] = [];
-    segments.forEach((segment) => {
-      let match = segment.match(/^:?([^?]+)\?$/);
-      if (match) {
-        optionalParams.push(match[1]);
-      }
-    });
-
-    if (optionalParams.length > 0) {
-      for (let i = 0; i <= optionalParams.length; i++) {
-        let newPath = path;
-        let newMeta = routesMeta.map((m) => ({ ...m }));
-
-        for (let j = optionalParams.length - 1; j >= 0; j--) {
-          let re = new RegExp(`(\\/:?${optionalParams[j]})\\?`);
-          let replacement = j < i ? "$1" : "";
-          newPath = newPath.replace(re, replacement);
-          newMeta[newMeta.length - 1].relativePath = newMeta[
-            newMeta.length - 1
-          ].relativePath.replace(re, replacement);
-        }
-
-        branches.push({
-          path: newPath,
-          score: computeScore(newPath, route.index),
-          routesMeta: newMeta,
-        });
-      }
-    } else {
+    if (!path.includes("?")) {
       branches.push({
         path,
         score: computeScore(path, route.index),
         routesMeta,
       });
+      return;
     }
+
+    explode(path).forEach((explodedPath) => {
+      let [explodedRoutesMeta, last] = tail(routesMeta);
+      if (last !== undefined) {
+        explodedRoutesMeta.push({
+          ...last,
+          relativePath: explodedPath.slice(parentPath.length),
+        });
+      }
+      branches.push({
+        path: explodedPath,
+        score: computeScore(explodedPath, route.index),
+        routesMeta: explodedRoutesMeta,
+      });
+    });
   });
 
   return branches;
 }
 
+function tail<Item>(arr: Item[]): [Item[], Item | undefined] {
+  return [arr.slice(0, -1), arr.at(-1)];
+}
+
+function explode(path: string): string[] {
+  let segments = path.split("/");
+  if (segments.length === 0) return [];
+
+  let [first, ...rest] = segments;
+  let isOptional = first.endsWith("?");
+  let segment = isOptional ? first.slice(0, -1) : first;
+
+  if (rest.length === 0) {
+    if (!isOptional) return [segment];
+    return ["", segment];
+  }
+
+  let restExploded = explode(rest.join("/"));
+  let exploded = restExploded.map((x) => segment + "/" + x);
+
+  if (!isOptional) return exploded;
+  return [...restExploded, ...exploded];
```

Comment:
I know this is WIP but took a quick look and I think we're getting some unintended flattened routes with dynamic params.  `/nested/:one?/:two?` is producing:

<img width="433" alt="Screenshot 2022-11-30 at 10 45 44 AM" src="https://user-images.githubusercontent.com/1609022/204843818-088dcc8b-841a-4001-a4c2-233b7392347e.png">

But I don't think we want `nested/:two` since the first param should always "win".  If we got rid of the trailing slash I think they'd have the same score and we could fool with the ordering to get `/nested/:one` first - but I think when it comes to optional dynamic param segments (unlike optional static segments) we should to take order into account.  I think this is what the `j < i` part was doing in the prior approach (but it was a long time ago that I wrote that so I can't be sure 😂 )

Reply:
I'll definitely add tests for that case since right now everything passes!

But yes, I can fix by adding a step to prune the exploded paths based on optional segment priority.